service/history/timer_queue_task_executor_base.go

```diff
@@ -200,25 +204,103 @@ func (t *timerQueueTaskExecutorBase) isValidExecutionTimeoutTask(
 	}
-	smt, err := def.Deserialize(cbt.Info.Data, hsm.TaskKindTimer{Deadline: cbt.VisibilityTimestamp})
+	smt, err := def.Deserialize(timer.Data, hsm.TaskKindTimer{Deadline: deadline})
 	if err != nil {
-		return hsm.Ref{}, nil, true, fmt.Errorf(
+		return fmt.Errorf(
 			"%w: %w",
-			queues.NewUnprocessableTaskError(fmt.Sprintf("cannot deserialize task %v", cbt.Info.Type)),
+			queues.NewUnprocessableTaskError(fmt.Sprintf("cannot deserialize task %v", timer.Type)),
 			err,
 		)
 	}
-	return hsm.Ref{
-		WorkflowKey:     taskWorkflowKey(task),
-		StateMachineRef: cbt.Info.Ref,
-	}, smt, true, nil
+	ref := hsm.Ref{
+		WorkflowKey:     ms.GetWorkflowKey(),
+		StateMachineRef: timer.Ref,
+	}
+	if err := t.validateStateMachineRef(ms, ref); err != nil {
+		return err
+	}
+	node, err := ms.HSM().Child(ref.StateMachinePath())
+	if err != nil {
+		return err
+	}
+
+	if err := execute(node, smt); err != nil {
+		return fmt.Errorf("failed to execute task: %w", err)
+	}
+	return nil
+}
+
+func (t *timerQueueTaskExecutorBase) executeStateMachineTimerTask(
+	ctx context.Context,
+	task *tasks.StateMachineTimerTask,
+	execute func(node *hsm.Node, task hsm.Task) error,
+) (retError error) {
+	ctx, cancel := context.WithTimeout(ctx, taskTimeout)
+	defer cancel()
+
+	wfCtx, release, ms, err := t.getValidatedMutableState(
+		ctx, task.WorkflowKey, func(ms workflow.MutableState) error {
+			return workflow.TransitionHistoryStalenessCheck(
+				ms.GetExecutionInfo().GetTransitionHistory(),
+				task.Version,
+				task.MutableStateTransitionCount,
+			)
+		},
+	)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	timers := ms.GetExecutionInfo().StateMachineTimers
+
+	// StateMachineTimers are sorted by Deadline, iterate through them as long as the deadline is expired.
+	for len(timers) > 0 {
+		group := timers[0]
+		if !queues.IsTimeExpired(t.shardContext.GetTimeSource().Now(), group.Deadline.AsTime()) {
+			break
+		}
+
+		for _, timer := range group.Infos {
+			err := t.executeSingleStateMachineTimer(ms, group.Deadline.AsTime(), timer, execute)
+			if err != nil {
+				if !errors.As(err, new(*serviceerror.NotFound)) {
+					// Return on first error as we don't want to duplicate the Executable's error handling logic.
+					// This implies that a single bad task in the mutable state timer sequence will cause all other
+					// tasks to be stuck. We'll accept this limitation for now.
+					return err
+				}
+				// TODO(bergundy): Metric?
+				t.logger.Warn("Skipped state machine timer", tag.Error(err))
+			}
+		}
+		// Remove the processed timer group.
+		timers = timers[1:]
+	}
+
+	// We haven't done any work, return without committing.
+	if len(timers) == len(ms.GetExecutionInfo().StateMachineTimers) {
+		return nil
+	}
+
+	// Update processed timers and ensure the next timer task is scheduled.
+	ms.GetExecutionInfo().StateMachineTimers = timers
+	workflow.AddNextStateMachineTimerTask(ms)
```

Comment:
this will be done when closing transaction?

Reply:
Oh yeah, I forgot this would be done automatically and I don't need to worry about it.

Reply:
Had to revert this change, it doesn't work well with how mutable state handles transition history ATM. I generally agree this should be done in task generator though.